LTI provider (#29)
pressbooks/docs: _pages/lti.md

* [Manually Set up a LTI Configuration](#manually-set-up-a-lti-configuration)
* [Common Cartridge](#common-cartridge)

This documentation is up to date as of version 1.1.2 of the Pressbooks LTI Provider plugin.

# Installation
Get the plugin here: [https://github.com/pressbooks/pressbooks-lti-provider](https://github.com/pressbooks/pressbooks-lti-provider)

> This plugin should be installed and activated on the network level, but has configurations available both at the network and book level.

Reviewer: ```suggestion
This plugin must be installed and activated on the network level, but has configurations available both at the network and book level.
```

<div style="background-color: #fff8e1;padding: 8px">Note: if the domain was whitelisted while the LTI configuration was initially (automatically) set up and the domain was subsequently removed from the whitelist, the LTI connection *will still work*.</div>
To disable an LTI connection, go to **Network admin > Integrations > LTI Consumers**, select the target connection, and either trash it or uncheck the "**Enabled**" box.

> **Note:** Automatic configuration is a feature of the LTI2.0 specification and is not widely supported by Learning Management Systems.

Author: I believe this is accurate, but would appreciate confirmation from someone more knowledgeable.

> **Note:** Automatic configuration is a feature of the LTI2.0 specification and is not widely supported by Learning Management Systems.

Reviewer: ```suggestion
**Note:** Automatic configuration is a feature of the [LTI 2.0](https://www.imsglobal.org/specs/ltiv2p0) specification and is not widely supported by Learning Management Systems.
```

- Try to match a Pressbooks user by LTI ID (Stored in user_meta table.)
- If no match, then try to match a Pressbooks user by email.
- If there's no match, then check if we should create a user (Anonymous Guest = No, Everything Else = Yes).  When creating a user: Username = email prefix,  email = see above, and the LTI ID will be stored in the user_meta table. A user can have more than one LTI ID (Example: Moodle, Sakai, Canvas, Blackboard all point to the same Book and we can match the user's email).
- If the user does not have rights to the book, and role is not Anonymous Guest, then add them to the book with appropriate role and log them in.

# Manually Set Up a LTI Configuration
1. Click the Add new button on the LTI Consumers page. You will be brought to the "Adding LTI Consumer" form:
<a href="https://pressbooks.org/app/uploads/sites/2/2018/06/pb-lti-addconsumer.png"><img class="alignnone size-large wp-image-521" src="https://pressbooks.org/app/uploads/sites/2/2018/06/pb-lti-addconsumer-1024x507.png" alt="" width="840" height="416" /></a>
2. Fill in:
- **Name**: Name of LTI consumer

> - **Key**: identification key used for the LTI protocol; can be any string (though some LMSes will reject a key if it is too long).

Author: @greatislander Dac and I discovered that one of the keys I created was causing the LTI registration to fail because the LMS considered it too long. I haven't been able to determine what the proper length is/should be, but did find [this post from 2012](http://swl10.blogspot.com/2012/05/ims-lti-and-length-of-oauthconsumerkey.html) (no idea if it's relevant or still a valid issue, though).

**Note**: once a configuration is created, it will not be possible to edit the "**Key**" and "**Secret**" fields.

# Configuring Pressbooks with a Tool Consumer

> While the steps for configuring LTI providers differ among different tool consumers (like Learning Management Systems), a typical registration process will require a key/secret pair and a launch URL. The key/secret pair can be obtained following the manual "Adding LTI Consumer" form instructions above. The launch URL can be either the URL for the network itself, i.e. [https://yourinstitution.pressbooks.pub] or the URL for a particular book, i.e. [https://yourinstitution.pressbooks/pub/mybook].

Reviewer: ```suggestion
While the steps for configuring LTI providers differ among different tool consumers (like Learning Management Systems), a typical registration process will require a key/secret pair and a launch URL. The key/secret pair can be obtained following the manual "Adding LTI Consumer" form instructions above. The launch URL can be either the URL for the network itself, i.e. [https://yourinstitution.pressbooks.pub] or the URL for a particular book, i.e. [https://yourinstitution.pressbooks.pub/mybook].
```
